### DISCUSSION TITLE
Signature is expired problem!

### DISCUSSION THREAD

**Discussion by saskokrstevski123** [openpgpjs/openpgpjs#1256] [Q&A]
We are using openpgp in our project in the last two years. Suddenly, we faced a problem with signature validation. The same signature with the same public key sometimes pass validation, sometimes don't. 

Validation error 
{
1201|api   |   keyid: Keyid { bytes: '?�¬ñ½�¿A' },
1201|api   |   verified: Promise {
1201|api   |     <rejected> Error: Signature is expired
1201|api   |         at /home/solecom/solenet-main-backend/node_modules/openpgp/dist/openpgp.js:36133:15
1201|api   |         at processTicksAndRejections (internal/process/task_queues.js:93:5)
1201|api   |   },
1201|api   |   signature: Signature { packets: Array { '0': [Signature], length: 1 } },
1201|api   |   valid: false,
1201|api   |   error: Error: Signature is expired
1201|api   |       at /home/solecom/solenet-main-backend/node_modules/openpgp/dist/openpgp.js:36133:15
1201|api   |       at processTicksAndRejections (internal/process/task_queues.js:93:5)
1201|api   | }

But if we create signature, and we pause the validation for 1 s or more, validation pass with no error

{
1201|api   |   keyid: Keyid { bytes: '?�¬ñ½�¿A' },
1201|api   |   verified: Promise { true },
1201|api   |   signature: Signature { packets: Array { '0': [Signature], length: 1 } },
1201|api   |   valid: true
1201|api   | }

What is a problem, and is there solution?

**Comment by twiss** [comment]
Hey :wave:
Probably this means that the signature appears to be created at a time in the future. Signature expiry is calculated as follows: https://github.com/openpgpjs/openpgpjs/blob/9bc2cf38761cd183f1bb7fb1d034883382152f5d/src/packet/signature.js#L765

If you are verifying the signature on a different computer than where you're generating it, and the latter has a local clock which is ahead, or the former has a local clock which is behind, you can run into time skew problems like this.

One way to fix this is to pass a `date` parameter. E.g., when generating:
```js
openpgp.sign({ date: new Date(Date.now() - 1000), ... })
```
(this will set the creation date of the signature as if it was created one second ago).
Note that if you're using a key that has just been generated, this will also not work, you have to then pass `date` to `openpgp.generateKey`, as well.

Or, when verifying:
```js
openpgp.verify({ date: new Date(Date.now() + 1000), ... })
```
(this will pretend that we're currently one second in the future).

Note that will also make keys and signatures expire one second earlier, if they do expire. That's probably not really a big deal, but it's good to keep in mind if you increase the amount of seconds a lot.
